pydis_site/apps/api/serializers.py

```diff
@@ -249,15 +256,69 @@ class Meta:
         fields = ('id', 'name', 'colour', 'permissions', 'position')
 
 
-class UserSerializer(BulkSerializerMixin, ModelSerializer):
+class UserListSerializer(ListSerializer):
+    """List serializer for User model to handle bulk updates."""
+
+    def create(self, validated_data: list) -> list:
+        """Override create method to optimize django queries."""
+        new_users = []
+        seen = set()
+
+        for user_dict in validated_data:
+            if user_dict["id"] in seen:
+                raise ValidationError(
+                    {"id": f"User with ID {user_dict['id']} given multiple times."}
+                )
+            seen.add(user_dict["id"])
+            new_users.append(User(**user_dict))
+
+        return User.objects.bulk_create(new_users, ignore_conflicts=True)
+
+    def update(self, instance: QuerySet, validated_data: list) -> list:
+        """
+        Override update method to support bulk updates.
+
+        ref:https://www.django-rest-framework.org/api-guide/serializers/#customizing-multiple-update
+        """
+        instance_mapping = {user.id: user for user in instance}
+
+        updated = []
+        fields_to_update = set()
+        for user_data in validated_data:
+            for key in user_data:
+                fields_to_update.add(key)
+
+                try:
+                    user = instance_mapping[user_data["id"]]
+                except KeyError:
+                    raise NotFound({"id": f"User with id {user_data['id']} not found."})
+
+                user.__dict__.update(user_data)
+                updated.append(user)
+
+        fields_to_update.remove("id")
+
+        if not fields_to_update:
+            # Raise ValidationError when only id field is given.
+            raise ValidationError({"data": "Insufficient data provided."})
```

Comment:
By convention, using `"data"` as a key name implies that there's something wrong with a field named `data`. If you need to raise `ValidationErrors` that aren't specific to a field, use the configured [`NON_FIELD_ERRORS_KEY`](https://www.django-rest-framework.org/api-guide/settings/#non_field_errors_key) as the key for consistency. 

Furthermore, maybe the value should be in a list since this seems to be DRF convention. I think the idea is that there could be several errors for the same field/key, hence the list.